core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java

```diff
@@ -0,0 +1,513 @@
+        int idx = Math.abs(Hashing.murmur3_32().hashString(entry.getKey()).asInt()) % tservers.size();
+        location = tservers.get(idx);
+      }
+
+      List<Range> merged = Range.mergeOverlapping(Lists.transform(entry.getValue(), tm -> tm.getExtent().toDataRange())); // merge contiguous ranges
+      List<TRowRange> ranges = merged.stream().map(r -> toClippedExtent(r).toThrift()).collect(Collectors.toList()); // clip ranges to queried range
+
+      locations.computeIfAbsent(location, s -> new HashMap<>()).put(entry.getKey(), ranges);
+    }
+
+    return locations;
+  }
+
+  private <K,V> Iterable<Map<K,V>> partition(Map<K,V> map, int max) {
+
+    if (map.size() < max) {
+      return Collections.singletonList(map);
+    }
+
+    return new Iterable<Map<K,V>>() {
+      @Override
+      public Iterator<Map<K,V>> iterator() {
+        Iterator<Entry<K,V>> esi = map.entrySet().iterator();
+
+        return new Iterator<Map<K,V>>() {
+          @Override
+          public boolean hasNext() {
+            return esi.hasNext();
+          }
+
+          @Override
+          public Map<K,V> next() {
+            Map<K,V> workingMap = new HashMap<>(max);
+            while (esi.hasNext() && workingMap.size() < max) {
+              Entry<K,V> entry = esi.next();
+              workingMap.put(entry.getKey(), entry.getValue());
+            }
+            return workingMap;
+          }
+        };
+      }
+    };
+  }
+
+  private static class ProcessedFiles {
+    final SummaryCollection summaries;
+    final Set<String> failedFiles;
+
+    public ProcessedFiles() {
+      this.summaries = new SummaryCollection();
+      this.failedFiles = new HashSet<>();
+    }
+  }
+
+  private class FilesProcessor implements Callable<ProcessedFiles> {
+
+    HostAndPort location;
+    Map<String,List<TRowRange>> allFiles;
+    private TInfo tinfo;
+
+    public FilesProcessor(TInfo tinfo, HostAndPort location, Map<String,List<TRowRange>> allFiles) {
+      this.location = location;
+      this.allFiles = allFiles;
+      this.tinfo = tinfo;
+    }
+
+    @Override
+    public ProcessedFiles call() throws Exception {
+      ProcessedFiles pfiles = new ProcessedFiles();
+
+      Client client = null;
+      try {
+        client = ThriftUtil.getTServerClient(location, ctx);
+        // partition files into smaller chunks so that not too many are sent to a tserver at once
+        for (Map<String,List<TRowRange>> files : partition(allFiles, 500)) {
+          if (pfiles.failedFiles.size() > 0) {
+            // there was a previous failure on this tserver, so just fail the rest of the files
+            pfiles.failedFiles.addAll(files.keySet());
+            continue;
+          }
+
+          TSummaries tSums = null;
+          try {
+            tSums = client.getSummariesFromFiles(tinfo, ctx.rpcCreds(), getRequest(), files);
```

Comment:
Should we have some kind of inspection as to this cross-sectional RPC? What kind of metrics/logging can we add to understand this process better?

If I want to get the summaries for a table as a client, and that table is hosted by every tserver, so that mean we might get a cartesian product of RPCs from tserver->tserver?